proposals/Scoped-Custom-Element-Registries.md

```diff
@@ -119,6 +119,51 @@ As a result, it must limit constructors by default to only looking up registrati
 
 This poses a limitation for authors trying to use the constructor to create new elements associated to scoped registries but not registered as global. More flexibility can be analyzed post MVP, for now, a user-land abstraction can help by keeping track of the constructor and its respective registry.
 
+## Interaction with Declarative Shadow DOM
+
+[Declarative shadow DOM](https://github.com/mfreed7/declarative-shadow-dom/blob/master/README.md) allows HTML to construct shadow roots for elements from `<template>` elements with a `shadowrootmode` attribute.
+
+Since these shadow roots are not created by a host calling `attachShadow()`, the host doesn't have a chance to pass in a scoped custom element registry. If a host is using a scoped registry, we need to force the declarative shadow root to not use the global registry and instead leave custom elements un-upgraded until the host can create and assign the correct registry.
+
+To do this we add a `shadowrootregistry` attribute, according to the [declarative shadow root explainer section on additional `attachShadow()` options](https://github.com/mfreed7/declarative-shadow-dom/blob/master/README.md#additional-arguments-for-attachshadow). This attribute causes the declarative shadow root to have no registry associated with it at all:
+
+```html
+<my-element>
+  <template shadowrootmode="open" shadowrootregistry="">
+    <some-scoped-element></some-scoped-element>
+  </template>
+</my-element>
+```
+
+The shadow root created by this HTML will have a `null` registry, and be in a "awaiting scoped registry" state that allows the registry to be set once after creation.
+
+To identify this "no registry, but awaiting one" state, ShadowRoot will have a `scopedRegistry` boolean property. `scopedRegistry` will set to `true` for all ShadowRoots with a scoped registry, or awaiting a scoped registry. Code can tell that ShadowRoot has an assignable `registry` property if `root.registry === null && root.scopedRegistry === true`.
```

Comment:
We should probably have "shadowRoot.customElements" instead of "shadowRoot.scopedRegistry" to be consistent with the global registry.

Reply:
right.

Reply:
Hmm, I find `shadowRoot.customElements` a bit harder to understand, as it's defined. In other words, this boolean is true "when the shadowRoot has a scoped registry, or is waiting for a scoped registry". I feel like "registry" should be at least somewhere in the name of the property, right?

Reply:
@rniwa meant that `registry` should be named `customElements` to match the global `window.customElements`.

`shadowRoot.scopedRegistry` would still be the boolean, if that's agreeable. (I don't think we do `is` or `has` prefixes in the DOM, correct?)

Reply:
Oh! Thanks, I missed that. I agree with renaming `shadowRoot.registry` to `shadowRoot.customElements`, which is exactly what you both wrote, but I failed to read correctly.

The DOM does use `is` sometimes, e.g. `isConnected`, `isTrusted`, `isMap`, `isContentEditable`, etc. The `has` prefix (which seems like what we'd want here) is also used, though less so. For example, `hasBeenActive`. If `has` is acceptable, I'd definitely lobby for renaming `shadowRoot.scopedRegistry` to `shadowRoot.hasScopedRegistry`.